packages/react-devtools-shared/src/hookNamesCache.js

```diff
@@ -110,44 +110,51 @@ export function loadHookNames(
 
     let didTimeout = false;
 
-    loadHookNamesFunction(hooksTree, fetchFileWithCaching).then(
-      function onSuccess(hookNames) {
-        if (didTimeout) {
-          return;
-        }
-
-        if (__DEBUG__) {
-          console.log('[hookNamesCache] onSuccess() hookNames:', hookNames);
-        }
-
-        if (hookNames) {
-          const resolvedRecord = ((newRecord: any): ResolvedRecord<HookNames>);
-          resolvedRecord.status = Resolved;
-          resolvedRecord.value = hookNames;
-        } else {
-          const notFoundRecord = ((newRecord: any): RejectedRecord);
-          notFoundRecord.status = Rejected;
-          notFoundRecord.value = null;
-        }
-
-        wake();
-      },
-      function onError(error) {
-        if (didTimeout) {
-          return;
-        }
+    const onSuccess = hookNames => {
+      if (didTimeout) {
+        return;
+      }
+
+      if (__DEBUG__) {
+        console.log('[hookNamesCache] onSuccess() hookNames:', hookNames);
+      }
 
-        if (__DEBUG__) {
-          console.log('[hookNamesCache] onError() error:', error);
-        }
+      if (hookNames) {
+        const resolvedRecord = ((newRecord: any): ResolvedRecord<HookNames>);
+        resolvedRecord.status = Resolved;
+        resolvedRecord.value = hookNames;
+      } else {
+        const notFoundRecord = ((newRecord: any): RejectedRecord);
+        notFoundRecord.status = Rejected;
+        notFoundRecord.value = null;
+      }
 
-        const thrownRecord = ((newRecord: any): RejectedRecord);
-        thrownRecord.status = Rejected;
-        thrownRecord.value = null;
+      wake();
+    };
 
-        wake();
-      },
-    );
+    const onError = error => {
+      if (didTimeout) {
+        return;
+      }
+
+      if (__DEBUG__) {
+        console.log('[hookNamesCache] onError()');
+      }
+
+      console.error(error);
+
+      const thrownRecord = ((newRecord: any): RejectedRecord);
+      thrownRecord.status = Rejected;
+      thrownRecord.value = null;
+
+      wake();
+    };
+
+    const thenable = loadHookNamesFunction(hooksTree, fetchFileWithCaching);
+    thenable.then(onSuccess, onError);
+    if (typeof (thenable: any).catch === 'function') {
```

Comment:
This does not look right to me. Can you elaborate on the reasoning for why this makes sense?

Normally, `catch` is not a part of the Thenable contract altogether. In Promises, `.catch(onError)` is a convenience that translates to `.then(undefined, onError)`. So from what I see in this implementation, if `thenable` is a Promise, the consequence should be that `onError` runs twice. I don't think this is what was intended.

Do you have a specific example case that you're trying to fix here? I suspect there's some other problem that was covered up, and maybe this accidentally fixes it, but it seems like a wrong fix.

Reply:
As I understand it, `catch` should really only be necessary if an error was thrown inside of `onSuccess` or `onError` – but what I noticed yesterday when testing a source-map parsing problem (inside of the `hooks/parseHookNames` module) was that the `onError` callback wasn't being invoked, and the error was effectively being swallowed (until the timeout eventually fired).

I can try to repro this by backing out this change, if that would be helpful.

Reply:
I checked out `bvaughn:devtools-dynamic-import-suspense-cache` (#22263), reverted this commit (24c2e27256b8f2051bde626598f2fba34efd109b), and tried to repro the problem I saw yesterday and I'm not able. Kind of a head scratcher, _but_ it tracks with my understanding of how `.catch()` works.

I can revert this specific change from this PR and leave the rest. Maybe @jstejada can weigh in if he sees another case where we aren't handling an error correctly in the cache?

Reply:
Reverted in https://github.com/facebook/react/pull/22263/commits/3c3f0722e0f3e9a855089f6672d8c47ac83d10cf

Reply:
Bit of follow-up.

After reverting that bit of the change, and using the test shell (with HMR on) for a little bit, I think I see the "uncaught error" issue I initially noticed. Digging in a little more, I think it actually comes from _this_ code which _doesn't pass the error handler_:

https://github.com/facebook/react/blob/8f96c6b2ac3ff0e38f60c87b1f08deb9993526b5/packages/react-devtools-shared/src/inspectedElementCache.js#L178-L195

So I think this was my mistake. Thanks for questioning this.

Reply:
not sure if related, but yes i am planning to look into a case where we seemed to not handle an error, will report back as to whether it's related to this or not

Reply:
> As I understand it, catch should really only be necessary if an error was thrown inside of onSuccess or onError

So I think it’s okay (and intentional) not to catch errors in these, just like React doesn’t. Because they should have the minimal logic that sets the values in the cache and have no reason to throw. If they do throw then that’s a bug in the cache, which _should_ be surfaced as unhandled.